File: logi-plugin-sdk-python/buffers/buffer_manager.py
class BufferManager:
    def __init__(self, client, ws):
        self.buffers = {}
        self.controlsRegistry = {}
        self.client = client
        self.handlers = {
            'UpdateAnalogControl': lambda msg: self.handleUpdateAnalogControl(msg),
            'ResponseInfo': lambda msg, ws: self.handleResponseInfo(msg, ws)
        }
    
    def isBufferingEnabled(self):
        return len(self.controlsRegistry) > 0
    
    def registerControls(self, controls):
        for controlId, bufferName in controls.items():
            self.controlsRegistry[controlId] = bufferName
    
    def registerBuffer(self, bufferName, buffer):
        buffer.setClient(self.client)
        self.buffers[bufferName] = buffer
    
    def processEvent(self, msgType, msgObj):
        if msgType not in self.handlers or not self.handlers[msgType](msgObj, self.ws):
            self.client.dispatchMessage(msgType, msgObj)
    
    def handleUpdateAnalogControl(self, msg):
        if msg['message']['analogControlId'] in self.controlsRegistry:
            bufferName = self.controlsRegistry[msg['message']['analogControlId']]
            if bufferName in self.buffers:
                self.buffers[bufferName].onUpdateAnalogControl(msg['id'], msg['message'])
                return True
        return False
    
    def handleResponseInfo(self, msg, ws):
        for _, buffer in self.buffers.items():
            buffer.onResponseInfo(msg['id'], ws)

File: sdk/buffers/buffer_manager.js
class BufferManager {
  #buffers = new Map();
  #controlsRegistry = new Map();
  #client;
  #handlers;
  #ws;

  constructor(client, ws) {
    this.#client = client;
    this.#ws = ws;

    this.#handlers = {
      ['UpdateAnalogControl']: (msg) => {
        return this.handleUpdateAnalogControl(msg);
      },
      ['ResponseInfo']: (msg, ws) => {
        return this.handleResponseInfo(msg, ws);
      },
    };
  }

  isBufferingEnabled() {
    return this.#controlsRegistry.size > 0;
  }

  registerControls(controls) {
    for (const [controlId, bufferName] of controls) {
      this.#controlsRegistry.set(controlId, bufferName);
    }
  }

  registerBuffer(bufferName, buffer) {
    buffer.setClient(this.#client);
    this.#buffers.set(bufferName, buffer);
  }

  processEvent(msgType, msgObj) {
    if (!this.#handlers.hasOwnProperty(msgType) || !this.#handlers[msgType](msgObj, this.#ws)) {
      this.#client.dispatchMessage(msgType, msgObj);
    }
  }

  handleUpdateAnalogControl({ id, message }) {
    if (this.#controlsRegistry.has(message.analogControlId)) {
      const bufferName = this.#controlsRegistry.get(message.analogControlId);
      if (this.#buffers.has(bufferName)) {
        this.#buffers.get(bufferName).onUpdateAnalogControl({ id, message });
        return true;
      }
    }

    return false;
  }

  handleResponseInfo({ id }, ws) {
    for (const [_, buffer] of this.#buffers) {
      buffer.onResponseInfo({ id }, ws);
    }

    return true;
  }

  flushBuffer (bufferName) {
    if (this.#buffers.has(bufferName)) {
      this.#buffers.get(bufferName).flushBuffer();
    }
  }

  flushAllBuffers () {
    for (const [_, buffer] of this.#buffers) {
      buffer.flushBuffer();
    }
  }

}

export default BufferManager;
